File: scripts/build-search-index.mjs
import { promises as fs } from 'node:fs';
import path from 'node:path';

const rootDir = path.resolve('./');
const guidesDir = path.join(rootDir, 'guides');
const dataDir = path.join(rootDir, 'data');
const outputFile = path.join(dataDir, 'search.json');

async function readGuideFiles() {
  const entries = await fs.readdir(guidesDir, { withFileTypes: true });
  return entries
    .filter((entry) => entry.isFile() && entry.name.endsWith('.html'))
    .map((entry) => entry.name)
    .sort((a, b) => a.localeCompare(b));
}

function extractTitle(contents) {
  const titleMatch = contents.match(/<title>([^<]*)<\/title>/i);
  if (!titleMatch) {
    return 'Untitled';
  }
  const raw = titleMatch[1].trim();
  const separatorIndex = raw.indexOf('|');
  if (separatorIndex === -1) {
    return raw;
  }
  return raw.slice(0, separatorIndex).trim();
}

function extractDescription(contents) {
  const descriptionMatch = contents.match(/<meta\s+name="description"\s+content="([^"]*)"/i);
  return descriptionMatch ? descriptionMatch[1].trim() : '';
}

function extractSections(contents) {
  const sections = [];
  const headingRegex = /<h2[^>]*>([\s\S]*?)<\/h2>/gi;
  let match;
  while (sections.length < 4 && (match = headingRegex.exec(contents))) {
    const text = match[1]
      .replace(/<[^>]+>/g, ' ')
      .replace(/&nbsp;/gi, ' ')
      .replace(/&amp;/gi, '&')
      .replace(/&rarr;/gi, '→')
      .replace(/\s+/g, ' ')
      .trim();
    if (text) {
      sections.push(text);
    }
  }
  return sections;
}

async function buildIndex() {
  const files = await readGuideFiles();
  const records = [];

  for (const file of files) {
    const filePath = path.join(guidesDir, file);
    const contents = await fs.readFile(filePath, 'utf8');

    records.push({
      title: extractTitle(contents),
      url: `guides/${file}`,
      abstract: extractDescription(contents),
      sections: extractSections(contents),
    });
  }

  return records;
}

async function main() {
  const index = await buildIndex();
  const pretty = `${JSON.stringify(index, null, 2)}\n`;
  await fs.writeFile(outputFile, pretty, 'utf8');
  console.log(`search.json rebuilt with ${index.length} entries.`);
}

main().catch((error) => {
  console.error('Failed to build search index', error);
  process.exitCode = 1;
});

File: scripts/build_search_index.py
import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
GUIDES_DIR = ROOT / "guides"
OUTPUT_FILE = ROOT / "data" / "search.json"
TOC_FILE = ROOT / "data" / "toc.json"
PROMPT_CATALOG = ROOT / "data" / "prompt-library" / "catalog.json"
TEMPLATE_CATALOG = ROOT / "data" / "templates" / "library.json"

H2_RE = re.compile(r"<h2[^>]*>([\s\S]*?)</h2>", re.IGNORECASE)
TITLE_RE = re.compile(r"<title>([^<]*)</title>", re.IGNORECASE)
DESC_RE = re.compile(r"<meta\s+name=\"description\"\s+content=\"([^\"]*)\"", re.IGNORECASE)
TAG_RE = re.compile(r"<[^>]+>")


def clean_text(raw: str) -> str:
    text = TAG_RE.sub(" ", raw)
    text = (text
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&rarr;", "→"))
    return " ".join(text.split())


def extract_title(contents: str) -> str:
    match = TITLE_RE.search(contents)
    if not match:
        return "Untitled"
    raw = match.group(1).strip()
    if "|" in raw:
        raw = raw.split("|", 1)[0].strip()
    return raw


def extract_description(contents: str) -> str:
    match = DESC_RE.search(contents)
    return match.group(1).strip() if match else ""


def extract_sections(contents: str, limit: int = 4) -> list[str]:
    sections: list[str] = []
    for match in H2_RE.finditer(contents):
        candidate = clean_text(match.group(1))
        if candidate:
            sections.append(candidate)
        if len(sections) >= limit:
            break
    return sections


def load_toc_map() -> dict[str, str]:
    if not TOC_FILE.exists():
        return {}
    toc_data = json.loads(TOC_FILE.read_text(encoding="utf-8"))
    mapping: dict[str, str] = {}
    for unit in toc_data:
        unit_title = unit.get("title", "")
        for item in unit.get("items", []):
            mapping[item.get("url", "")] = unit_title
    return mapping


def build_guide_records(toc_map: dict[str, str]) -> list[dict]:
    records: list[dict] = []
    for html_file in sorted(GUIDES_DIR.glob("*.html")):
        contents = html_file.read_text(encoding="utf-8")
        url = f"guides/{html_file.name}"
        title = extract_title(contents)
        abstract = extract_description(contents)
        sections = extract_sections(contents)
        unit = toc_map.get(url, "Guide")
        meta = f"Guide | {unit}" if unit else "Guide"

        records.append({
            "type": "guide",
            "title": title,
            "url": url,
            "unit": unit,
            "meta": meta,
            "snippet": abstract or " ".join(sections[:2]),
            "tags": sections,
        })
    return records


def build_prompt_records() -> list[dict]:
    if not PROMPT_CATALOG.exists():
        return []
    entries = json.loads(PROMPT_CATALOG.read_text(encoding="utf-8"))
    records: list[dict] = []
    for entry in entries:
        tags = entry.get("tags", [])[:]
        tags.extend([entry.get("category", ""), entry.get("difficulty", ""), entry.get("risk_level", "")])
        meta = "Prompt | {category} | {difficulty}".format(
            category=entry.get("category", "Uncategorised"),
            difficulty=entry.get("difficulty", ""),
        ).strip(" |")
        records.append({
            "type": "prompt",
            "title": f"{entry['id']} - {entry['title']}",
            "url": f"guides/84_prompt-library.html#{entry['id']}",
            "unit": entry.get("category", "Prompt"),
            "meta": meta,
            "snippet": entry.get("use_case", ""),
            "tags": [tag for tag in tags if tag],
            "category": entry.get("category"),
            "difficulty": entry.get("difficulty"),
        })
    return records


def build_template_records() -> list[dict]:
    if not TEMPLATE_CATALOG.exists():
        return []
    entries = json.loads(TEMPLATE_CATALOG.read_text(encoding="utf-8"))
    records: list[dict] = []
    for entry in entries:
        tags = entry.get("tags", [])[:]
        tags.extend([entry.get("category", ""), entry.get("format", ""), entry.get("difficulty", "")])
        meta = "Template | {category} | {format}".format(
            category=entry.get("category", "Uncategorised"),
            format=entry.get("format", ""),
        ).strip(" |")
        records.append({
            "type": "template",
            "title": f"{entry['id']} - {entry['title']}",
            "url": f"guides/80_templates-library.html#{entry['id']}",
            "unit": entry.get("category", "Template"),
            "meta": meta,
            "snippet": entry.get("summary", ""),
            "tags": [tag for tag in tags if tag],
            "category": entry.get("category"),
            "difficulty": entry.get("difficulty"),
            "format": entry.get("format"),
        })
    return records


def main() -> None:
    toc_map = load_toc_map()
    guide_records = build_guide_records(toc_map)
    prompt_records = build_prompt_records()
    template_records = build_template_records()

    index = guide_records + prompt_records + template_records
    OUTPUT_FILE.write_text(json.dumps(index, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(
        "search.json rebuilt with "
        f"{len(guide_records)} guides, {len(prompt_records)} prompts, "
        f"{len(template_records)} templates"
    )


if __name__ == "__main__":
    main()
